[Inline][WinEH] Fix try scopes leaking to caller on inline

MuellerMP opened the pull request:
This fixes issue #164169

When inlining functions compiled with -EHa, try scope terminators might need to be inserted before inlined returns. 
This prevents leaking try scopes over to the caller.

efriedma-quic commented on llvm/lib/Transforms/Utils/InlineFunction.cpp:
```diff
@@ -2210,6 +2211,135 @@ inlineRetainOrClaimRVCalls(CallBase &CB, objcarc::ARCInstKind RVCallKind,
+    auto *BB = ReturnBlock->splitBasicBlockBefore(ReturnBlock->getTerminator(),
+                                                  "try.end");
+    BB->getTerminator()->eraseFromParent();
+    IRBuilder<>(BB).CreateInvoke(SehTryEndFn, ReturnBlock, UnwindDestination);
```
The strategy of inserting an seh.try.end is better than nothing, I guess.  But I think it's also worth looking at clang to see if we're accidentally skipping the emission of seh.try.end in some cases; it's much easier to reason about the scopes if they're properly balanced when they're first emitted.

efriedma-quic replied:
The "ret" instruction in the assembly doesn't actually need to be inside the try block for compatibility, as far as I can tell.  Both MSVC and LLVM are perfectly happy to emit the epilogue outside of the try block.  Trying to ensure the LLVM IR ret instruction comes before the seh.try.end is not at all necessary.

For other code motion... we currently don't strictly enforce code motion in a try block with EHa.  Trying to reconcile the notion of a trap with the way LLVM usually reasons about code is very hard; usually there's no restrictions on generated code if it triggers undefined behavior.  LLVM tries to catch exceptions that aren't explicitly raised using RaiseException etc., but only to the extent that it can without interfering with the way optimization normally works.  There's probably room for improvement, but -EHa is a pretty niche feature; nobody has really been interested in investing time to improve it beyond what's already there.

efriedma-quic reviewed (commented):
You can rebase/merge main to force a rebuild; arm64 bots should be fixed now, I think.

Please investigate why clang is emitting ret instructions without emitting an appropriate seh.try.end first.

MuellerMP commented on llvm/lib/Transforms/Utils/InlineFunction.cpp:
```diff
@@ -2210,6 +2211,135 @@ inlineRetainOrClaimRVCalls(CallBase &CB, objcarc::ARCInstKind RVCallKind,
+
+      auto InsertIt = Scopes.end();
+      if (!Scopes.empty())
+        for (auto ScopeIt = Scopes.begin(); ScopeIt != Scopes.end(); ++ScopeIt)
```
removed this loop
should perform better now

MuellerMP committed:
[Inline][WinEH] Fix try scopes leaking to caller on inline

This fixes issue #164169

When inlining functions compiled with -EHa, try scope terminators might need to be inserted before inlined returns.
This prevents leaking try scopes over to the caller.

MuellerMP commented on llvm/lib/Transforms/Utils/InlineFunction.cpp:
```diff
@@ -2210,6 +2211,135 @@ inlineRetainOrClaimRVCalls(CallBase &CB, objcarc::ARCInstKind RVCallKind,
+      continue;
+
+    // If there is a catch which connects to the return, the try scope is
+    // considered to be terminated.
```
seh.try.end invokes are now also integrated if they exist.
I also added a seh specific test which on one hand makes sure that we find the seh.try.end terminator and on the other hand also asserts a correct termination for seh.try.begin invokes for a inlined callee with non cxx personality.

MuellerMP commented:
Thanks for your help @efriedma-quic - very much appreciated!
As you pointed out it probably makes more sense to try and fix this in the frontend as this will also make the handling less complex.

Will close this and come back if I have a better fix.
